### DISCUSSION TITLE
Migrating to 3.x. Do not ignore MainActivity and AppDelegate callback changes

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1398] [Q&A] [answered]
When moving to Shiny 3.x you will find that in Android this code

```
public override void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Android.Content.PM.Permission[] grantResults)
{
    this.ShinyOnRequestPermissionsResult(requestCode, permissions, grantResults);
    base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
}
```

will no longer compile because the ShinyOnRequestPermissionsResult extension has been replaced. Do not naively think "oh Shiny 3.x must be hooking this internally now". That call to ShinyOnRequestPermissionsResult needs to be replaced with a call to `Host.Lifecycle.OnRequestPermissionsResult(this, requestCode, permissions, grantResults);`

If you do not do this then your RequestAccess calls will never return. The permission will be granted (assuming the user agreed) and subsequent calls will return but the one that triggered OnRequestPermissionsResult will not.

I spent far too long looking elsewhere for the problem suspecting threading deadlocks. Don't be me :-)

The ShinyAndroidActivityclass in 3.x does have this in it base implementation and it is how I realised my mistake, https://github.com/shinyorg/shiny/blob/dev/src/Shiny.Hosting.Native/Platforms/Android/ShinyAndroidActivity.cs#L38

There is a similar change in the iOS' AppDelegate for 

- RegisteredForRemoteNotifications
- DidReceiveRemoteNotification
- FailedToRegisterForRemoteNotifications
- HandleEventsForBackgroundUrl

**Comment by aritchie** [comment, marked as answer]
Hey @munkii - take a look at this: https://shinylib.net/client/hosting/manual/

I think this covers everything

**Reply by munkii** [reply]
Looks good to me. We had one other overload that I could not see a Host equivalent for. I don't believe we were relying on this as it is part of the background jobs stuff and we don't use that.

`public override void PerformFetch(UIApplication application, Action<UIBackgroundFetchResult> completionHandler) => this.ShinyPerformFetch(completionHandler);
`

FWIW My AppDelegate FinishedLaunching now looks like this.

      
```
/// <summary>
/// Finished launching.
/// </summary>
/// <param name="app">The application.</param>
/// <param name="options">The options.</param>
/// <returns>A bool.</returns>
public override bool FinishedLaunching(UIApplication app, NSDictionary options)
{
    string appCenterKey = null;

#if !RELEASE_LIVECONFIG
    appCenterKey = AppSettings.AppCenterAnalyticsIosDevTest;
#else
    appCenterKey = AppSettings.AppCenterAnalyticsIos;
#endif

    var host = HostBuilder.Create();
    host
        .Services
        .AddBluetoothLE<BleDelegate>()
        .AddPush<PushDelegate>()
        .AddNotifications<NotificationDelegate>()
        .RegisterCommonServices()
        .RegisterPlatformServices();

    host.
        Logging
        .AddAppCenter(appCenterKey, Microsoft.Extensions.Logging.LogLevel.Error);

    host
        .Build()
        .Run();

    return base.FinishedLaunching(app, options);
}
```

**Reply by aritchie** [reply]
The host builder will register these things on its own

        .RegisterCommonServices()
        .RegisterPlatformServices();

**Reply by munkii** [reply]
Those two methods are mine.

RegisterCommonServices registers our common singleton dependencies to IServiceCollection 
RegisterPlatformServices contains what used to be in 

```
Our.ShinyStartup shinyStartup = new Our.ShinyStartup()
{
    RegisterPlatformServices = (services) =>
    {

    },
};
```
Our you saying there is a Shiny3.0 pair of methods we should be plugging into?

**Reply by aritchie** [reply]
No you're good.  I have methods named similar and I thought you were calling into those.

You don't need the appfetch stuff for shiny jobs any longer for people that do use them.
